src/app/pipes/invalid-and-interaction.pipe.spec.ts

```diff
@@ -0,0 +1,45 @@
+import { FormControl, Validators } from '@angular/forms';
+import { InvalidAndInteractionPipe } from './invalid-and-interaction.pipe';
+
+describe('InvalidAndInteractionPipe', () => {
+  it('create an instance', () => {
+    const pipe = new InvalidAndInteractionPipe();
+    expect(pipe).toBeTruthy();
+  });
+
+  it('should return false if the control is valid and there has been no interaction', () => {
+    const formControl = new FormControl('hello', Validators.required);
+    const pipe = new InvalidAndInteractionPipe();
+
+    const transformed = pipe.transform(formControl);
+    expect(transformed).toBeFalse();
+  });
+
+  it('should return false if the control is valid and there has been interaction', () => {
+    const formControl = new FormControl('hello', Validators.required);
+    const pipe = new InvalidAndInteractionPipe();
+    formControl.markAsDirty();
+    formControl.markAsTouched();
+
+    const transformed = pipe.transform(formControl);
+    expect(transformed).toBeFalse();
+  });
+
+  it('should return false if the control is invalid and there has been no interaction', () => {
+    const formControl = new FormControl('', Validators.required);
+    const pipe = new InvalidAndInteractionPipe();
+
+    const transformed = pipe.transform(formControl);
+    expect(transformed).toBeFalse();
+  });
+
+  it('should return true if the form is invalid and there has been interaction', () => {
```

Comment:
Minor point but to cover everything, the dirty and touched states should be tested in isolation.
Also don't know if it's worth testing against some stub/mock AbstractControl instead of a concrete FormControl, so you're not relying on its behaviour.

Reply:
I've separated those cases and replaced the form controls with (partial) jasmine mocks of abstract controls. I added a few complementary properties to the ones used in the implementation of the pipe to try to cover a few possible implementation changes. I also parameterised the tests.